rust_src/src/buffers.rs

```diff
@@ -1843,4 +1847,81 @@ pub fn byte_char_debug_check(b: LispBufferRef, charpos: isize, bytepos: isize) {
     }
 }
 
+// True if B can be used as 'other-than-BUFFER' buffer.
+fn candidate_buffer(b: LispObject, buffer: LispObject) -> bool {
+    match b.as_buffer() {
+        Some(buf) => !b.eq(buffer) && buf.is_live() && !buf.is_hidden(),
+        None => false,
+    }
+}
+
+/// Return most recently selected buffer other than BUFFER.
+/// Buffers not visible in windows are preferred to visible buffers, unless
+/// optional second argument VISIBLE-OK is non-nil.  Ignore the argument
+/// BUFFER unless it denotes a live buffer.  If the optional third argument
+/// FRAME specifies a live frame, then use that frame's buffer list instead
+/// of the selected frame's buffer list.
+///
+/// The buffer is found by scanning the selected or specified frame's buffer
+/// list first, followed by the list of all buffers.  If no other buffer
+/// exists, return the buffer `*scratch*' (creating it if necessary).
+#[lisp_fn(min = "0")]
+pub fn other_buffer(buffer: LispObject, visible_ok: LispObject, frame: LispObject) -> LispObject {
+    let f: LispFrameRef = LispFrameLiveOrSelected::from(frame).into();
+    let pred = f.buffer_predicate;
+    let mut notsogood = None;
+
+    let frame_bufs = f
+        .buffer_list
+        .iter_cars(LispConsEndChecks::off, LispConsCircularChecks::off);
+    let all_bufs = LiveBufferIter::new().map(LispObject::from);
+    for buf in frame_bufs.chain(all_bufs) {
+        // If the frame has a buffer_predicate, disregard buffers that
+        // don't fit the predicate.
+        if candidate_buffer(buf, buffer)
+            && pred.map_or_else(|| true, |p| call!(p, buf).is_not_nil())
+        {
+            if visible_ok.is_not_nil() || unsafe { Fget_buffer_window(buf, Qvisible) }.is_nil() {
+                return buf;
+            } else {
+                notsogood = notsogood.or(Some(buf));
+            }
+        }
+    }
+
+    match notsogood {
+        Some(buf) => buf,
+        None => {
+            // TODO: This was AUTO_STRING, which doesn't exist yet in Rust.
+            let scratch = new_unibyte_string!("*scratch*");
+            let mut buf = Fget_buffer(scratch);
+            if buf.is_nil() {
+                buf = Fget_buffer_create(scratch);
+                unsafe { Fset_buffer_major_mode(buf) };
+            }
+            buf
+        }
+    }
+}
+
+// The following function is a safe variant of Fother_buffer: It doesn't
+// pay attention to any frame-local buffer lists, doesn't care about
+// visibility of buffers, and doesn't evaluate any frame predicates.
+#[no_mangle]
+pub extern "C" fn other_buffer_safely(buffer: LispObject) -> LispObject {
+    for buf in LiveBufferIter::new().map(|x| x.into()) {
+        if candidate_buffer(buf, buffer) {
+            return buf;
+        }
+    }
+
```

Comment:
This block looks like a shared function.

Reply:
Oh, yes. Thanks.